### PR TITLE
Fix symbolic controlled-U qubit allocation

### PR DESCRIPTION
**Pull request by ksk-jij**
## Summary

Fix symbolic controlled-U resource allocation so scalar controls that are actually qubit-array elements alias their existing physical wires instead of allocating phantom qubits.

## Details

- Pass emit-time bindings into `ResourceAllocator._allocate_controlled_u` so symbolic array-element controls such as `q[control_qubit]` can resolve through the existing root-array qubit map.
- Map multi-argument symbolic controlled-U control results from the resolved physical operand groups instead of relying on scalar UUID aliases created by the allocator.
- Add regression coverage for the reported 6-to-14 Qiskit width inflation, literal-index symbolic control counts, loop-indexed controls, then-use control results, and optional QURI Parts/CUDA-Q width checks.

## Verification

- `uv run pytest tests/transpiler/backends/test_symbolic_controlled_width.py -q`
- `uv run ruff check qamomile/ tests/`
- `uv run ruff format --check qamomile/ tests/`
- `uv run zuban check qamomile/`
- `uv run pytest tests/transpiler/ -q`
- `uv run pytest -q`

### CODE REVIEW

File: tests/transpiler/backends/test_symbolic_controlled_width.py [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+    assert _cudaq_width(kernel, bindings) == expected_width
+
+
+def test_scalar_input_symbolic_control_allocator_preserves_fresh_result_alias():
+    """Fresh scalar controls still map their results for later use."""
+    block = scalar_input_symbolic_control_then_use.build(parameters=["num_controls"])
+    qubit_map, _ = ResourceAllocator(ValueResolver()).allocate(
+        block.operations,
+        bindings={"num_controls": 2},
+    )
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply tests/transpiler/backends/test_symbolic_controlled_width.py]: Fixed in c2f4635d: the allocator-only scalar-control test now keeps num_controls as a runtime parameter without also passing it as a compile-time binding.

**Comment by ksk-jij** [comment]
# Branch Summary

## 1. Problem Overview

This branch fixes a bug where the transpiler allocated non-existent extra qubits and inflated backend circuit width when a scalar qubit taken from a `Vector`, such as `q[control_qubit]`, was passed as a control-prefix operand to `qmc.control(...)` inside a Qamomile qkernel. On main, when a multi-argument `SymbolicControlledU` (a controlled-U IR operation whose control count or operand width is determined by compile-time bindings) included a scalar control, `ResourceAllocator` could not find that scalar UUID in `qubit_map` and treated it like a fresh top-level scalar input instead of an existing Vector element, allocating a new physical wire. This branch passes active bindings into the allocator and, when a scalar operand has a parent `ArrayValue` and element index, resolves the original Vector element's physical wire through `ValueResolver.resolve_qubit_index(...)`; as a result, repeated symbolic MCZ operations over a 6-qubit register still emit a 6-qubit circuit.

One important IR detail: after a gate or controlled-U operation, Qamomile creates a new `Value` / UUID, but the quantum state is not copied to a different qubit. In many cases the new value still points to the same physical backend wire. For example, in `q2 = qmc.h(q1)`, `q1` and `q2` are different IR values, but they represent the value before and after an H gate on the same backend wire. This summary calls such a result a pass-through result: a new result value whose physical wire is the same as its input. The fix is mostly about correctly updating `qubit_map` / `clbit_map`, the maps from IR `QubitAddress` keys to backend physical qubit / classical-bit indices, so the new UUID also points to the same physical index.

## 2. Frontend-Level Behavior

There is no user-facing API change. The behavior changes when a symbolic index and symbolic control count are resolved through `bindings` during transpilation: previously, `q[control_qubit]` could be counted as a phantom qubit; now, if a compile-time binding such as `control_qubit=5` is provided, it is treated as the original `q[5]` wire. Representative kernels are the symbolic-index MCZ in `tests/transpiler/backends/test_symbolic_controlled_width.py:22` and the repeated symbolic MCZ in `tests/transpiler/backends/test_symbolic_controlled_width.py:54`; the Qiskit regression at `tests/transpiler/backends/test_symbolic_controlled_width.py:369` checks that `circuit.num_qubits == 6`.

```python
import qamomile.circuit as qmc
from qamomile.qiskit import QiskitTranspiler

@qmc.qkernel
def kernel(total: qmc.UInt, control_qubit: qmc.UInt, signal_start: qmc.UInt, num_signal: qmc.UInt) -> qmc.Vector[qmc.Bit]:
    q = qmc.qubit_array(total, name="q")
    target = signal_start + num_signal - 1
    mcz = qmc.control(qmc.z, num_controls=num_signal)
    q[control_qubit], q[signal_start:target], q[target] = mcz(q[control_qubit], q[signal_start:target], q[target])
    return qmc.measure(q)

circuit = QiskitTranspiler().transpile(kernel, bindings={"total": 6, "control_qubit": 5, "signal_start": 3, "num_signal": 2}).compiled_quantum[0].circuit
assert circuit.num_qubits == 6
```

If the symbolic scalar index is left unbound, the transpiler does not silently treat it as a fresh qubit; it fails explicitly with `EmitError` during emission. This is covered at `tests/transpiler/backends/test_symbolic_controlled_width.py:406` and makes the existing contract clearer: an index that selects a qubit wire must be resolvable at transpile time.

## 3. Backend, IR, and Transpiler Changes

`ResourceAllocator` estimates physical qubit and classical-bit indices before backend emission. This branch passes active bindings into `ControlledUOperation` allocation (`qamomile/circuit/transpiler/passes/emit_support/resource_allocator.py:328`). `_allocate_controlled_u` now accepts `bindings` (`qamomile/circuit/transpiler/passes/emit_support/resource_allocator.py:724`), and when a `SymbolicControlledU` scalar control has `parent_array` and `element_indices`, it resolves the existing physical index of the parent Vector element through `ValueResolver.resolve_qubit_index(...)` (`qamomile/circuit/transpiler/passes/emit_support/resource_allocator.py:775`). When resolution succeeds, both the scalar control operand UUID and the corresponding result UUID are aliased to the same physical wire (`qamomile/circuit/transpiler/passes/emit_support/resource_allocator.py:780`, `qamomile/circuit/transpiler/passes/emit_support/resource_allocator.py:800`); only genuinely fresh top-level scalar inputs continue to receive a new allocation (`qamomile/circuit/transpiler/passes/emit_support/resource_allocator.py:785`).

ArrayValue physical-index map mutation is centralized in `qamomile/circuit/transpiler/passes/emit_support/physical_index_map.py:1`. `map_array_result_group(...)` registers result element addresses and a base address for a result UUID (`qamomile/circuit/transpiler/passes/emit_support/physical_index_map.py:14`). `array_element_mapping(...)` and `array_element_mappings(...)` read element-to-physical mappings for existing arrays (`qamomile/circuit/transpiler/passes/emit_support/physical_index_map.py:36`, `qamomile/circuit/transpiler/passes/emit_support/physical_index_map.py:59`). `map_array_element_aliases(...)` and `copy_array_element_aliases(...)` register pass-through results so they point to the same physical resources as their source arrays (`qamomile/circuit/transpiler/passes/emit_support/physical_index_map.py:84`, `qamomile/circuit/transpiler/passes/emit_support/physical_index_map.py:124`). This gives allocator, controlled emission, inverse emission, and phi remapping a shared element/base-address convention.

`emit_controlled_u_multi_arg` now preserves the multi-argument `SymbolicControlledU` control prefix as per-operand `control_index_groups`, instead of only flattening it (`qamomile/circuit/transpiler/passes/emit_support/controlled_emission.py:1477`). The backend gate emission path still uses flat `control_phys` (`qamomile/circuit/transpiler/passes/emit_support/controlled_emission.py:1482`), while result-side map updates pass the grouped operands into `_map_operand_result_groups(...)`, so scalar controls and Vector/slice controls remain aligned with their corresponding result operands (`qamomile/circuit/transpiler/passes/emit_support/controlled_emission.py:1547`). ArrayValue result mapping is routed through `map_array_result_group(...)` in the controlled emission paths so base-address handling is consistent (`qamomile/circuit/transpiler/passes/emit_support/controlled_emission.py:432`, `qamomile/circuit/transpiler/passes/emit_support/controlled_emission.py:1406`, `qamomile/circuit/transpiler/passes/emit_support/controlled_emission.py:1562`, `qamomile/circuit/transpiler/passes/emit_support/controlled_emission.py:2951`).

Phi outputs and inverse-block results now use the same helper family for map updates. `remap_static_phi_outputs` copies quantum ArrayValue phi outputs through `copy_array_element_aliases(...)` for compile-time constant branches (`qamomile/circuit/transpiler/passes/emit_support/condition_resolution.py:157`). `map_phi_outputs` reads true/false quantum ArrayValue mappings in a single scan with `array_element_mappings(...)`, and only registers output aliases when both branches point to the same physical resources (`qamomile/circuit/transpiler/passes/emit_support/condition_resolution.py:200`). Bit ArrayValue phi outputs register element aliases without a base address so classical-array results are not misread as scalar outputs (`qamomile/circuit/transpiler/passes/emit_support/condition_resolution.py:273`). Inverse-block ArrayValue results also use `map_array_result_group(...)` (`qamomile/circuit/transpiler/passes/emit_support/inverse_emission.py:28`, `qamomile/circuit/transpiler/passes/emit_support/inverse_emission.py:395`).

The regression tests are collected in `tests/transpiler/backends/test_symbolic_controlled_width.py:1`. Qiskit coverage checks repeated symbolic MCZ, literal index plus symbolic control count, subsequent use of the control result, loop-indexed scalar controls, the unresolved symbolic-index error path, neighboring non-offender paths, and controlled `modular_decrement` (`tests/transpiler/backends/test_symbolic_controlled_width.py:369`, `tests/transpiler/backends/test_symbolic_controlled_width.py:375`, `tests/transpiler/backends/test_symbolic_controlled_width.py:384`, `tests/transpiler/backends/test_symbolic_controlled_width.py:397`, `tests/transpiler/backends/test_symbolic_controlled_width.py:406`, `tests/transpiler/backends/test_symbolic_controlled_width.py:430`, `tests/transpiler/backends/test_symbolic_controlled_width.py:440`). Optional backend tests verify that the shared allocator fix also applies to QURI Parts and CUDA-Q (`tests/transpiler/backends/test_symbolic_controlled_width.py:450`, `tests/transpiler/backends/test_symbolic_controlled_width.py:468`). A direct `ResourceAllocator` test keeps the true scalar-input edge case covered, because fresh scalar controls still need allocation and result aliasing (`tests/transpiler/backends/test_symbolic_controlled_width.py:486`).

## 4. Alternatives Considered

One alternative was to lower multi-argument `SymbolicControlledU` into `ConcreteControlledU` earlier, or to pre-expand Vector/slice controls into a list of scalar controls at the IR level. This was not chosen because Qamomile's IR is designed to preserve what the program means and defer backend-specific concretization to emission. Losing the original multi-argument control grouping would make allocator and emit-side result mapping more fragile. The chosen approach keeps the IR shape intact, uses `ValueResolver` for allocation-time address resolution, and preserves operand grouping during emission before passing it into `_map_operand_result_groups(...)`.

Another alternative was to add backend-specific workarounds in the Qiskit, QURI Parts, and CUDA-Q emitters. That was also rejected because the width inflation is caused by shared resource allocation and symbolic multi-argument controlled emission map updates, not by a backend-specific gate API. Fixing `resource_allocator.py` and `controlled_emission.py` has a smaller blast radius and the optional backend tests confirm that the same shared fix applies across backends.

A third alternative was to keep ArrayValue alias logic handwritten at each call site. That would leave `condition_resolution.py`, `controlled_emission.py`, `inverse_emission.py`, and `resource_allocator.py` each responsible for element addresses and base addresses, making future changes to alias rules easy to miss in one path. This branch centralizes the mutation helpers in `physical_index_map.py`: `map_array_result_group(...)` handles result mapping, while `map_array_element_aliases(...)` and `copy_array_element_aliases(...)` handle pass-through results that inherit existing physical resources.

## 5. Known Limitations

When: `control_qubit` in `q[control_qubit]` is not resolvable from transpile-time `bindings`. Why: backend emission does not have a mechanism for selecting a qubit wire from a runtime-variable index, and `ValueResolver.resolve_qubit_index(...)` can only lower root-array elements that are known at compile time. Future fix: supporting runtime-indexed quantum operands would require a deliberate dynamic-qubit-addressing design in both the IR and backend emitters, rather than allocation-time aliasing.

When: a future operation outside `SymbolicControlledU`, Pauli evolve, composite/inverse, and phi outputs introduces a new quantum path that writes Vector-derived scalar operands or ArrayValue operands into pass-through results. Why: `physical_index_map.py` provides shared helpers, but a new operation will not register result-side mappings unless its allocator / emit path calls those helpers. Future fix: when adding such an operation, either use the operand-group-preserving result mapping helper or add the `ValueResolver`-based scalar-array-element aliasing logic to that operation's allocator / emit path.
